File: Java/Dwumiany.java
import java.util.HashMap;
import java.util.Scanner;

class Main 
{
    static long coefficient(int n, int k, HashMap<String, Long> factorialCache) 
    {
        if (n == k || k == 0)
            return 1;
        if (n == 0)
            return 0;
        String key = n + "," + k;
        if (factorialCache.containsKey(key))
            return factorialCache.get(key);
        long result = 1;
        for (int i = 1; i <= Math.min(k, n - k); i++) 
        {
            result *= n - i + 1;
            result /= i;
        }
        factorialCache.put(key, result);
        return result;
    }

    public static void main(String[] args) 
    {
        Scanner input = new Scanner(System.in);
        int T = input.nextInt();
        HashMap<String, Long> factorialCache = new HashMap<>();
        for (int t = 0; t < T; t++) 
        {
            int n = input.nextInt();
            int k = input.nextInt();
            System.out.println(coefficient(n, k, factorialCache));
        }
        input.close();
    }
}

File: Python/Dwumiany.py
def coefficient(n, k, factorial_cache):
    if n == k or k == 0:
        return 1
    if n == 0:
        return 0
    if (n, k) in factorial_cache:
        return factorial_cache[(n, k)]
    result = 1
    for i in range(1, min(k, n - k) + 1):
        result *= n - i + 1
        result //= i
    factorial_cache[(n, k)] = result
    return result

T = int(input())
factorial_cache = {}
for _ in range(T):
    n, k = map(int, input().split())
    print(coefficient(n, k, factorial_cache))
